crates/red_knot_python_semantic/src/types.rs

```diff
@@ -824,6 +823,66 @@ impl<'db> From<&Type<'db>> for Type<'db> {
     }
 }
 
+/// Non-exhaustive enumeration of builtin types to allow for easier syntax when interacting with
+/// the most common builtin types (e.g. int, str, ...).
+///
+/// Feel free to expend this enum if you ever find yourself using the same builtin type in multiple
+/// places.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum BuiltinType {
```

Comment:
Naming nit: let's rename this to `BuiltinClass`. This actually only handles classes, not all kinds of types (functions are types too).

Speaking of which, I kinda want to unify this PR and `FunctionKind`, in terms of both naming and API. The one trick there is that we want to represent known functions that aren't actually builtins (e.g. `reveal_type`). The same may be true in future for other stdlib classes!

So maybe ultimately what we'll actually want is `KnownClass` and `KnownFunction` enums, and a `known` field (which is an `Option<KnownClass>` or `Option<KnownFunction>` - I like this Option approach better than `FunctionKind::Ordinary`), and `is_known` method, on both `ClassType` and `FunctionType`. But this would require generalizing what you have in this PR to not assume builtins module, and instead classify the known functions/classes by module as well. I would be happy with doing all of this now, in this PR, or waiting on it if you prefer to land a simpler version of this PR for now.

cc @AlexWaygood in case you hate my naming/API preferences :)

Reply:
Hmm, I considered the pros and cons of `Option<KnownFunction>` vs having `FunctionKind::Ordinary` while working on my previous PR. I ended up going with `FunctionKindOrdinary` because I have a general preference having a flat list of possible states rather than representing possible states using an enum of enums. It makes it easier to count exactly how many possible states there are, and giving the default state a name (`Ordinary`) rather than using `None` makes it abundantly clear to new readers of the code what the default state represents. It's also more ergonomic to pattern-match on a flat list rather than a nested enum.

I don't feel _too_ strongly, but a flat enum would be my preference :)

Reply:
Yes, you're definitely right that `FunctionKind` is not a great name. And I think you're right that renaming it to `KnownFunction` addresses some of the concern I had that it wouldn't be obvious what the default state signified. It's pretty obvious that `None` indicates that it's not a known function!

I'm still not a massive fan of nested enums, but I guess in this case it's _okay_. Feel free to proceed :)

Reply:
Would you like me to address this naming change in this PR?

Reply:
Yes, if you're up for it I think it would be ideal to go ahead and make the switch to both `KnownClass` and `KnownFunction` enums in this PR, with `ClassType` having a field `known: Option<KnownClass>` and a method `is_known`, and similar for `FunctionType`. Where `KnownFunction` would replace the current `FunctionKind` enum, and the new `FunctionType::known` field would replace the current `FunctionType::kind`.

Reply:
The newest commit implements a first attempt at this. I did three things:
- Apply naming convention for `KnownFunction`
- Apply naming convention for `KnownClass`
- Try to extend the structure in `KnownClass` to easily support non-builtins